### ISSUE TITLE
malloc(): unsorted double linked list corrupted

### ISSUE THREAD

**Issue by erulabs** [imgproxy/imgproxy#1190]
Hello!

  This morning, one of our new imgproxy pro instances suddenly began timing out (hitting its `IMGPROXY_WRITE_TIMEOUT` value). This continued to occurred until we killed the container, even though healthchecks continued responding properly.

  Just before all requests began timing out, we saw an error log like:
  
> malloc(): unsorted double linked list corrupted

  This is on `imgproxy-pro:v3.18.2`. Some of the options provided:

```
        DD_RUNTIME_METRICS_ENABLED = "true"
        IMGPROXY_DATADOG_ENABLE_ADDITIONAL_METRICS = "true"
        DD_ENV                  = "prod"
        DD_SERVICE              = "imgproxy"
        ENV                     = "Prod"
        IMGPROXY_LOG_FORMAT     = "json"
        IMGPROXY_LOG_LEVEL      = "warn"
        IMGPROXY_DATADOG_ENABLE = "true"
        IMGPROXY_QUALITY        = "75"
        IMGPROXY_USE_S3         = "true"
        IMGPROXY_S3_REGION      = "us-east-1"
        IMGPROXY_STRIP_METADATA = "true"
        IMGPROXY_USE_ETAG       = "true"
        IMGPROXY_DOWNLOAD_TIMEOUT = "20"
        IMGPROXY_READ_TIMEOUT = "18"
        IMGPROXY_WRITE_TIMEOUT = "12"
        IMGPROXY_MAX_SRC_RESOLUTION = "500.0"
        IMGPROXY_MAX_SRC_FILE_SIZE = "100000000"
        IMGPROXY_SKIP_PROCESSING_FORMATS = "gif"
        IMGPROXY_MAX_ANIMATION_FRAMES = "500"
        IMGPROXY_MAX_ANIMATION_FRAME_RESOLUTION = "1000.0"
        IMGPROXY_CONCURRENCY    = "4"
```

  Note that we're working to compress our older and extremely large images and GIFs, so ideally we can remove/lower some of these super high limits shortly.

  Unfortunately, we do not know exactly what image caused this error (still trying to piece that together). We're only using "resize:fill" and plain (ie: no resize provided) image compression.

  In this situation, it might be better if imgproxy either crashes or returns failures via the healthcheck.

  Please let me know if I can provide more information, I apologize for how lite on info this issue is - Ill update when/if I can find more details!

Thank you!

**Comment by DarthSim** [comment]
Hi @erulabs,

The only suspect I have here is the resizing and convolution code. But since you're saying that resizing is not involved, that seems to not be the case. One of the images that came right before requests began timing out is probably the guilty one. Having this image and the processing options you've used will help to debug this.

**Comment by DarthSim** [comment]
Do you by any chance save some images to WebP?

**Comment by erulabs** [comment]
@DarthSim No, we're currently outputting to jpeg. To clarify, resizing is involved, but we only use the following URI patterns currently:

   /0/plain/s3:/%[path]
   /0/page:%[url_param(p)]/plain/s3:/%[path]
   /0/resize:fill::%[url_param(h)]/plain/s3:/%[path]

Ill enable info-level logging so we can get an idea of what image caused this. I'd guess resize:fill is the one the failed based on your comment!

**Comment by DarthSim** [comment]
`v3.19.0` includes some changes that may fix this issue. Also, I changed the `/health` endpoint so it checks that vips is operational. This should make health checks fail in case of such issues.

**Comment by erulabs** [comment]
Excellent, the /health change is perfect. I only had a bit of time to look thru the code, but I wonder if this error was a side-effect of having a lower IMGPROXY_WRITE_TIMEOUT than IMGPROXY_READ_TIMEOUT. Is it correct to assume READ_TIMEOUT should be greater than or equal to WRITE_TIMEOUT?

Rolling out 3.19.0 now, thank you!

**Comment by DarthSim** [comment]
`IMGPROXY_READ_TIMEOUT` is the time limit for reading the request from the client. Since imgproxy only accepts GET requests, there's no need to increase this config value. `IMGPROXY_READ_TIMEOUT` and `IMGPROXY_WRITE_TIMEOUT` do not conflict as they affect different stages of request processing. And they definitely couldn't cause this issue.

Stuff like this could only happen on the C side as this issue is caused by work with raw memory (probably, buffer overflow or some kind of such stuff). Probably, this was caused by one of the patches we apply to vips in imgproxy distribution. As I said before, I have a suspect, but according to my calculations, it shouldn't cause the problem. Also, I made lots of tests before releasing the version that uses it, and after you reported the issue, but no issues were detected. Anyway, I made some changes that should prevent problems with that patch even if my calculations were wrong.
